doozerlib/image.py

```diff
@@ -191,3 +212,237 @@ def get_additional_push_names(self, additional_registries):
+                # See examples: https://gist.github.com/jupierce/adeb7b2b10f5d225c8090bab80640011
+                tag = koji_api.getTag(tag_name)
+                if not tag:
+                    # Some sort of pseudo tag like rhel-8.0.0-z-batch-0.3-set
+                    continue
+                tag_perm = tag.get('perm', None)
+
+                if not tag_perm:
+                    # Looks to be someone else's candidate tag or other non-shipment related tag.
+                    # rhaos-4.4-rhel-7 has tag_perm=='trusted'
+                    # kpatch-kernel-4.18.0-193.6.3.el8_2-build  has tag_perm='admin'
+                    continue
+
+                # tag_name now represents indicates another product shipped this build. Through various
+                # methods, our image build could have pick it up from this location. We aren't going to
+                # try to figure it out for certain -- just add it to the list of possibilities.
+                relevant_tags.add(tag_name)
+
+            if not relevant_tags:
+                # Just a sanity check
+                raise IOError(f'Found no relevant tags which make {rpm_package_name} available to {dgk}')
+
+            # Now we have a list of tags that were conceivably the source of our image build's RPM.
+            for rel_tag_name in relevant_tags:
+                # Let's determine if a later package has been tagged with this relevant tag.
+                # Example output: https://gist.github.com/jupierce/0e4a0d1e97b1a764fb5b7af08429a75c
+                latest_tagged = koji_api.getLatestRPMS(tag=rel_tag_name, package=rpm_package_name)[1]
+                build_event_ids = set([rel_build['creation_event_id'] for rel_build in latest_tagged])
+                # Record this list of events against the package build id so we don't have to do this
+                # calculation again (many images likely use similar rpm package builds).
+                with older_than_lock:
+                    # This is not perfect -- another thread may also have computed this for another
+                    # image, but we don't want to hold the lock for that long. The race condition is
+                    # safe to allow.
+                    ImageMetadata.package_build_ids_build_events_cache[package_build_id] = build_event_ids
+
+        # We have a list of build events that might have affected what we will pull in from tags
+        # which have previously been applied to a package build we pulled into the latest image.
+        # See if those builds happened more recently than our latest image. If they have, when we
+        # rebuild, those tags might contribute the new RPM to the new image. If they don't that is
+        # also fine. The new latest image will have a creation_event_id newer than the build events,
+        # and we will not retrigger this logic.
+        for rel_build_event_id in build_event_ids:
+            if rel_build_event_id > image_build_event_id:
```

Comment:
Haven't thought this much, but it seems to me that it is possible that if someone untags an RPM from a tag, the latest RPM could change to an older version. So i think we need to request a rebuild if the version-releases are different from those in the image.

Reply:
A gap, but a difficult one to close. For example, it is perfectly valid for our image to contain version Y of an RPM, while version X  (X<Y) is in our candidate tag. This could be a scenario like you describe above (rare) or one in which we shipped something in the past but now rely on another product to provide it (somewhat less rare). 
I'm generally OK with leaving this as a forced rebuild, but we should document the limitation.

Reply:
yeah, it should be rare. maybe we can also mention some known limitations and points that can be improved in the future in TODO or FIXME comments.